### ISSUE TITLE
Issue 3: Add fair non-GNN and GNN baselines under identical splits

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/12-graph-based-cyber-physical-risk#24]
# Issue 3: Add fair non-GNN and GNN baselines under identical splits

## Milestone: M2 — Does the GNN Actually Add Value?

---

## Problem Statement

Compare GraphSAGE against strong non-graph and alternative graph baselines using **exactly the same data, feature sets, train/validation/test samples, threshold-selection rules, and metrics**.

The purpose is not to declare GraphSAGE the winner. The purpose is to determine whether a graph neural network contributes measurable value beyond ordinary tabular learning.

---

## Research Question Supported

**RQ2:** Under leakage-safe evaluation, how much predictive value comes from genuine graph structure compared with packet/flow content?

---

### 1. Non-GNN Baselines

At minimum implement:

1. **Logistic Regression**
2. **MLP**
3. **One strong tree-based model**
   - XGBoost if available and easy to reproduce, otherwise
   - Random Forest / HistGradientBoosting.

### 2. GNN Baselines

Implement:

4. **GCN**
5. **GraphSAGE** — current reference model
6. **GAT**

Keep architectures modest and comparable.

Do not spend excessive effort hyperparameter-tuning every model.

### 3. Fairness Requirements

For each experiment:

- same train/validation/test IDs,
- same raw feature selection,
- same encoded content features where applicable,
- validation-only threshold selection,
- same evaluation metrics,
- same random-seed set where stochastic,
- no model may see test labels during tuning.

If a non-GNN model cannot consume topology, it should receive the **same per-flow content features** used by the GraphSAGE edge classifier.

If degree features are supplied explicitly, document them as a separate feature condition rather than mixing them silently into the content baseline.

### 4. Required Experiments

Run the baseline comparison under at least these protocols:

#### Experiment B1 — SCADANet Track B

Compare all six model families on the same Track-B split.

This answers:

> For known attacker behavior across broad attack coverage, does GraphSAGE outperform ordinary classifiers?

#### Experiment B2 — Strict SCADANet temporal evaluation

After Issue #5 is complete, compare the principal baseline models on the strict chronological protocol.

At minimum:

- tree model,
- MLP,
- GraphSAGE,
- best alternative GNN from B1.

#### Experiment B3 — BATADAL strict temporal evaluation

Compare:

- MLP or temporal-window tabular baseline,
- GraphSAGE,
- GCN or GAT.

For the tabular BATADAL baseline, flatten or aggregate the window features in a clearly documented manner.

#### Optional Experiment B4 — ICS-Flow

Use only if it adds insight. Do not force all six models onto ICS-Flow if the small topology makes the experiment scientifically weak.

### 5. Hyperparameter Policy

Use a small validation-only search.

For example:

```text
hidden dimension: 32, 64
learning rate: 1e-3, 5e-3
dropout: 0.0, 0.2
```

Tree models may use a similarly limited search.

Report the search space in the repository.

Do not use the test set to select hyperparameters.

### 6. Required Metrics

For every model report:

- Precision
- Recall
- F1
- AUPRC
- FPR
- AUROC if meaningful
- inference latency
- number of trainable parameters for neural models
- training runtime where practical.

Also report **per-attack recall** on SCADANet.

### 7. Result Artifacts

Create:

```text
experiments/results/q1/baselines/
```

Suggested artifacts:

```text
scadanet_trackb_baselines.csv
scadanet_temporal_baselines.csv
batadal_temporal_baselines.csv
baseline_configs.json
```

Save per-seed results rather than only the final mean.

#### Table B1 — Main baseline comparison

| Dataset | Protocol | Model | Precision | Recall | F1 | AUPRC | FPR |
|---|---|---|---:|---:|---:|---:|---:|
| SCADANet | Track B | Logistic Regression | | | | | |
| SCADANet | Track B | Tree model | | | | | |
| SCADANet | Track B | MLP | | | | | |
| SCADANet | Track B | GCN | | | | | |
| SCADANet | Track B | GraphSAGE | | | | | |
| SCADANet | Track B | GAT | | | | | |

#### Table B2 — Temporal baseline comparison

| Dataset | Model | Precision | Recall | F1 | AUPRC | FPR | ΔF1 vs static |
|---|---|---:|---:|---:|---:|---:|---:|

#### Table B3 — Computational cost

| Model | Parameters | Train time | Inference time | Peak memory if available |
|---|---:|---:|---:|---:|

#### Figure B1 — Model performance under the same protocol

Grouped bar plot or point plot showing:

- F1
- AUPRC
- FPR

for each model on SCADANet Track B.

Avoid using accuracy as the main visual.

#### Figure B2 — Static/known-behavior vs temporal performance

For the principal models:

- x-axis = model,
- two values = non-temporal vs strict temporal F1,
- optionally a second panel/file for AUPRC.

#### Figure B3 — Per-attack recall heatmap

Rows = SCADANet attack types  
Columns = selected models  
Cells = recall.

This is important because aggregate F1 may hide attack classes that completely fail.

---

## Interpretation Rules

The following outcomes are all scientifically useful:

### Outcome 1

```text
GraphSAGE > MLP/tree
```

and later Issue #4 shows the advantage disappears with shuffled topology.

Interpretation: genuine graph relationships help.

### Outcome 2

```text
GraphSAGE ≈ MLP/tree
```

Interpretation: the graph architecture provides little additional predictive value under the tested protocol.

### Outcome 3

```text
GraphSAGE < tree/MLP
```

Interpretation: the benchmark does not justify GNN complexity; this strengthens the paper's critique of assuming that graph structure automatically improves ICS detection.

Do not hide an unfavorable outcome.

---

## Acceptance Criteria

- [ ] At least three non-GNN/GNN model families are compared fairly, with the preferred full set listed above.
- [ ] Identical split IDs are used across models.
- [ ] Hyperparameter selection uses validation data only.
- [ ] All configurations are committed.
- [ ] Per-seed results are saved.
- [ ] Table B1 and B2 can be generated from saved results.
- [ ] Figure B1–B3 are generated automatically.
- [ ] Per-attack recall is reported for SCADANet.
- [ ] Computational cost is reported for the principal models.
- [ ] Conclusions explicitly state whether the GNN provides value, rather than assuming it.

---

## Do Not Do In This Issue

Do not:

- invent a new GNN architecture,
- perform an enormous hyperparameter sweep,
- use different feature sets to make one model look better,
- select models based on test-set performance,
- remove difficult attack classes.

This issue must produce a fair baseline answer, not a leaderboard.
